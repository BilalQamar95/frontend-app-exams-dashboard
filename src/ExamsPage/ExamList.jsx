import React from 'react';
import PropTypes from 'prop-types';

export const ExamList = ({ exams }) => (
  <div>
    <ul>
      {exams.map((exam) => (<li>{exam.name}</li>))}g
    </ul>
  </div>
);

ExamList.propTypes = {
  exams: PropTypes.arrayOf(PropTypes.object).isRequired, // eslint-disable-line react/forbid-prop-types
};

export default ExamList;
