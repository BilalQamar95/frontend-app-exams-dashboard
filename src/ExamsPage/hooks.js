import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RequestKeys } from 'data/constants';

import * as reduxHooks from 'data/redux/hooks';

import * as api from './data/api';
import { loadExams } from './data/reducer';

export const useFetchCourseExams = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (courseId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchCourseExams,
      promise: api.getCourseExams(courseId),
      onSuccess: (response) => dispatch(loadExams(response)),
    })
  );
};

export const useInitializeExamsPage = (courseId) => {
  const fetchCourseExams = useFetchCourseExams();
  React.useEffect(() => { fetchCourseExams(courseId); }, []); // eslint-disable-line react-hooks/exhaustive-deps
};

export const useExamsLoading = () => (
  reduxHooks.useRequestIsPending(RequestKeys.fetchCourseExams)
);

export const useCourseExamsList = () => (
  useSelector(state => state.exams.examsList)
);
