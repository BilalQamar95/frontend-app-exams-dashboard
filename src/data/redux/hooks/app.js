// App wide selectors would be implemented here if we had any.