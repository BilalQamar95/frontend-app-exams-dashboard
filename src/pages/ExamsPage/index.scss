@import "~@edx/paragon/scss/core/core";

#exam-selector {
  margin: 0.5em auto;
}

.pgn__searchfield .form-control {
  border: 1px solid $search-border-color !important;
}

.pgn__searchfield.has-focus:not(.pgn__searchfield--external)::after {
  border: none;
  width: 0px;
}
