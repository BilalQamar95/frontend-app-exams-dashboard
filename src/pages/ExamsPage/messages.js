import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  attemptsViewTabTitle: {
    id: 'ExamsPage.attemptsViewTabTitle',
    defaultMessage: 'Attempts',
    description: 'Title for the attempts view tab',
  },
  reviewDashboardTabTitle: {
    id: 'ExamsPage.reviewDashboardTabTitle',
    defaultMessage: 'Review Dashboard',
    description: 'Title for the review dashboard tab',
  },
});

export default messages;
