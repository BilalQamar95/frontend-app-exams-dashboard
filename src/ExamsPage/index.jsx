import React, {
  useEffect
} from 'react';
import PropTypes from 'prop-types';

import { useCourseExamsList, useFetchCourseExams } from './hooks';


function ExamsPage({ courseId }) {
  const fetchExams = useFetchCourseExams();
  const examsList = useCourseExamsList();

  useEffect(() => {
    fetchExams(courseId);
  }, [courseId]);

  return (
    <div>
      <ul>
        {examsList.map((exam) => (<li>{exam.exam_name}</li>))}
      </ul>
    </div>
  );
}

ExamsPage.propTypes = {
  courseId: PropTypes.string.isRequired,
};

export default ExamsPage;
