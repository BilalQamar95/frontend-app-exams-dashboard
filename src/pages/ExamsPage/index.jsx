import React from 'react';
import PropTypes from 'prop-types';

import { useExamsData, useInitializeExamsPage } from './hooks';
import ExamList from './components/ExamList';

const ExamsPage = ({ courseId }) => {
  useInitializeExamsPage(courseId);
  const {
    examsList,
    isLoading,
    exampleValue,
    setExampleValue,
  } = useExamsData();

  return (
    <div>
      { isLoading && <div>Loading...</div>}
      <ExamList exams={examsList} />
      {/* count button just to demo attaching component state */}
      <button type="button" onClick={() => setExampleValue(exampleValue + 1)}>
        {exampleValue}
      </button>
    </div>
  );
};

ExamsPage.propTypes = {
  courseId: PropTypes.string.isRequired,
};

export default ExamsPage;
