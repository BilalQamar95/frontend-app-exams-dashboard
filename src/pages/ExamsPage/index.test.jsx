import { render } from '@testing-library/react';

import ExamsPage from '.';
import * as hooks from './hooks';

jest.mock('./hooks', () => ({
  useInitializeExamsPage: jest.fn(),
  useExamsData: jest.fn(),
}));

jest.mock('./components/ExamList', () => 'ExamList');

describe('ExamsPage', () => {
  const defaultExamsData = {
    examsList: 'stub-exams-list', // explicity testable in snapshots
    isLoading: false,
  };
  test('snapshot loading', () => {
    hooks.useExamsData.mockReturnValue({
      ...defaultExamsData,
      isLoading: true,
    });
    // TODO: this is rendering the ExamList as examlist and causing DOM errors on the
    // snapshot. Need to figure out why this is happening.
    expect(render(<ExamsPage courseId="test_course" />)).toMatchSnapshot();
  });
  test('snapshot loaded', () => {
    hooks.useExamsData.mockReturnValue(defaultExamsData);
    expect(render(<ExamsPage courseId="test_course" />)).toMatchSnapshot();
  });
});
