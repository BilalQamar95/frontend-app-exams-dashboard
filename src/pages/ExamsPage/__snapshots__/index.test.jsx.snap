// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExamsPage snapshots loaded 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-fluid"
      >
        <div>
          <ul>
            <li>
              exam1
            </li>
          </ul>
        </div>
        <div>
          <nav
            class="pgn__tabs nav nav-tabs"
            role="tablist"
          >
            <a
              aria-selected="true"
              class="nav-item nav-link active"
              data-rb-event-key="attempts"
              href="#"
              role="tab"
            >
              Attempts
            </a>
            <a
              aria-selected="false"
              class="nav-item nav-link"
              data-rb-event-key="review"
              href="#"
              role="tab"
            >
              Review Dashboard
            </a>
            <a
              class="pgn__tab_invisible pgn__tab_more nav-item nav-link"
              href="#"
              role="tab"
              tabindex="-1"
            >
              <div
                class="pgn__dropdown pgn__dropdown-light dropdown"
                data-testid="dropdown"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="nav-link dropdown-toggle btn btn-link"
                  id="pgn__tab-toggle"
                  type="button"
                >
                  More...
                </button>
              </div>
            </a>
          </nav>
          <div
            class="tab-content"
          >
            <div
              aria-hidden="false"
              class="fade tab-pane active show"
              role="tabpanel"
            >
              <div
                data-testid="attempt_list"
              >
                Placeholder for attempt list
              </div>
            </div>
            <div
              aria-hidden="true"
              class="fade tab-pane"
              role="tabpanel"
            >
              <div
                data-testid="review_dash"
              >
                Placeholder for external review app
              </div>
            </div>
            <div
              aria-hidden="true"
              class="fade tab-pane"
              role="tabpanel"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-fluid"
    >
      <div>
        <ul>
          <li>
            exam1
          </li>
        </ul>
      </div>
      <div>
        <nav
          class="pgn__tabs nav nav-tabs"
          role="tablist"
        >
          <a
            aria-selected="true"
            class="nav-item nav-link active"
            data-rb-event-key="attempts"
            href="#"
            role="tab"
          >
            Attempts
          </a>
          <a
            aria-selected="false"
            class="nav-item nav-link"
            data-rb-event-key="review"
            href="#"
            role="tab"
          >
            Review Dashboard
          </a>
          <a
            class="pgn__tab_invisible pgn__tab_more nav-item nav-link"
            href="#"
            role="tab"
            tabindex="-1"
          >
            <div
              class="pgn__dropdown pgn__dropdown-light dropdown"
              data-testid="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="nav-link dropdown-toggle btn btn-link"
                id="pgn__tab-toggle"
                type="button"
              >
                More...
              </button>
            </div>
          </a>
        </nav>
        <div
          class="tab-content"
        >
          <div
            aria-hidden="false"
            class="fade tab-pane active show"
            role="tabpanel"
          >
            <div
              data-testid="attempt_list"
            >
              Placeholder for attempt list
            </div>
          </div>
          <div
            aria-hidden="true"
            class="fade tab-pane"
            role="tabpanel"
          >
            <div
              data-testid="review_dash"
            >
              Placeholder for external review app
            </div>
          </div>
          <div
            aria-hidden="true"
            class="fade tab-pane"
            role="tabpanel"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
